File: programme/python/pol.py
from lib.CrcMoose import *
import serial
import getopt, sys, os
import struct, math

class Plugwise:
    def __init__(self, port, macaddress):
        self.serial = serial.Serial(port, "115200")
        
        self.HEADER = '\x05\x05\x03\x03'
        self.ENDLINE = '\x0d\x0a'
        self.CALIBRATIONCODE = '0026'
        self.POWERCHANGECODE = '0017'
        self.POWERINFOCODE = '0012'
        self.CALIBRATIONRESPONSECODE = '0027'
        self.POWERINFORESPONSECODE = '0013'
        
        self.macaddress = macaddress
        self.gain_a = 0
        self.gain_b = 0
        self.offtot = 0
        self.offruis = 0
        self.correctedpulses = 0

    def GetCalibrationInfo(self):
        self.SendCommand("0026" +  self.macaddress)
        result = self.GetResult(self.CALIBRATIONRESPONSECODE)
        
        self.gain_a = self.hexToFloat(result[0:8])
        self.gain_b = self.hexToFloat(result[8:16])
        self.offtot = self.hexToFloat(result[16:24])
        self.offruis = self.hexToFloat(result[24:32])
        
    def PulseCorrection(self, pulses):
        value = pulses / 1.0;
        out = 1.0 * (((pow(value + self.offruis, 2.0) * self.gain_b) + ((value + self.offruis) * self.gain_a)) + self.offtot)
        return out

    def pulseToKWH(self, pulses):
        output = (pulses / 1) / 468.9385193;
        return output

    def pulseToWatt(self, pulses):
        result = self.pulseToKWH(pulses) * 1000
        return result

    def GetPowerInfo(self):
        self.GetCalibrationInfo()
        self.SendCommand(self.POWERINFOCODE + self.macaddress)
        result = self.GetResult(self.POWERINFORESPONSECODE)
        resultint = self.hexToInt(result[4:8])
        corrected = self.PulseCorrection(resultint)

        return self.pulseToWatt(resultint)

    def SetPowerState(self, newstate):
        self.SendCommand(self.POWERCHANGECODE + self.macaddress + newstate)
        self.serial.readline()

    def GetCRC16(self, value):
        value = CRC16X.calcString(value)
        format = ("%%0%dX" % ((CRC16X.width + 3) // 4))
        return format % value

    def hexToFloat(self, hexstr):
        intval = int(hexstr, 16)
        bits = struct.pack('L', intval)
        return struct.unpack('f', bits)[0]

    def hexToInt(self, hexstr):
        return int(hexstr, 16)

    def SendCommand(self, command):
        self.serial.write(self.HEADER + command + self.GetCRC16(command) + \
        self.ENDLINE )

    def GetResult(self, responsecode):
        readbytes = 0
        
        if responsecode == self.CALIBRATIONRESPONSECODE:
            readbytes = 52
        elif responsecode == self.POWERINFORESPONSECODE:
            readbytes = 37
        elif responsecode == "0000":
            readbytes = 0

        data = ''
        
        while 1:
            data += self.serial.read(1)
            if data.endswith(responsecode):
                data = self.serial.read(readbytes)
                return data[16:]

def main():
    try:
        opts, args = getopt.getopt(sys.argv[1:], "p:o:f:w:h",
                     ['port', 'on', 'off', 'help', 'watt'])
    except getopt.error, why:
        print_help()

    port = 16
    macaddress = ''
    newstate = ''
    command = ''

    for o, a in opts:
        if (o in ('-p', '--port')):
            port = a
        if (o in ('-o', '--on')):
            newstate = '01'
            macaddress = a
            command = "ON"
        if (o in ('-f', '--off')):
            newstate = '00'
            macaddress = a
            command = "OFF"
        if (o in ('-w', '--watt')):
            macaddress = a
            command = "POWERINFO"
        if (o in ('-h', '--help')):
            print_help()

    if port == '' or macaddress == '':
        print_help()
    else:
        plugwise = Plugwise(port, macaddress)
        if command == "ON":
            plugwise.SetPowerState("01")
        elif command == "OFF":
            plugwise.SetPowerState("00")
        elif command == "POWERINFO":
            print plugwise.GetPowerInfo()
        
def print_help():
    print
    print "POL (Plugwise on Linux) v0.2 / Maarten Damen"
    print
    print "Usage: pol <options>"
    print
    print "Options:"
    print "  -p  --port <port>  Serial port of the Plugwise stick, for example: /dev/ttyUSB0"
    print "  -o  --on <macaddress>  Power on plugwise device with specified MAC address"
    print "  -f  --off <macaddress>  Power off plugwise device with specified MAC address"
    print "  -w  --watt <macaddress>  Get current watt usage for device with specified MAC address"
    print "  -h  --help   This help text"
    exit(0)

main()

File: dynamic_library/debug/pol.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include "crc.h"
#include "pol.h"

static const char HEADER[] = {5, 5, 3, 3, 0};
const char ENDLINE[] = {0x0d, 0x0a, 0};

const char POWERINFOCODE[] = "0012";
const char POWERINFORESPONSECODE[] = "0013";

static int fd;
static void init(void);

static void init(void)
{
  static int initDone = 0;
  if (initDone == 0){
    // BUG avec port, il faut le mettre en argument  ? static char * ?
    char *port ="/dev/ttyUSB0";
    
    fd = -1;
    while(fd == -1){
      printf("fd=%d\n",fd);
      fd = open(port, O_RDWR);
    }
    
    initDone = 1;
  }
}

float pulseToKWH(int pulses)
{
  return((pulses/1.0)/468.9385193);
}

float pulseToWatt(int pulses)
{
  return(pulseToKWH(pulses)*1000);
}

float getPowerInfo(Plugwise *pl)
{
  float corrected;
  char *result;
  int resultint;
  init();
  printf("init ok\n");
sendCommand(POWERINFOCODE, pl->macaddress); 
  result=getResult(pl , POWERINFORESPONSECODE);
  resultint=hexToInt(&result[4],4);
  return(pulseToWatt(resultint));
}

char *getCRC16(char *command)
{
  char result[5];
  unsigned short Checksum = calcString(command);
  sprintf(result, "%04X", Checksum);
  return(result);
}

int hexToInt(char *hexstr, int len)
{
  char tmp[64];
  strncpy(tmp, hexstr,len);
  tmp[len] = '\0';
  return strtoul(tmp, NULL, 16);
}

void sendCommand(const char *code, char *macAd)
{
  char command[128];
  char toSend[128];
  strcpy(command, code);
  strcat(command, macAd);
  strcpy(toSend, HEADER);
  strcat(toSend, command);
  strcat(toSend, getCRC16(command));
  strcat(toSend, ENDLINE);
  printf("to send:%s\n",toSend);
  serialWrite(toSend);
  printf("to send ok\n");
}

char *getResult(Plugwise *pl, char *responsecode)
{
  int readbytes = 0;
  int alreadyRead;
  char data[1024];
  int readReturn;
  
  if (responsecode == POWERINFORESPONSECODE)
    readbytes = 37;
  else if (responsecode=="0000")
    readbytes = 0;
  
  alreadyRead = 0;
  memset(data, '\0', sizeof(data));
  do{
    readReturn = -1;
    while (readReturn == -1){
      // or do{}while (readReturn != 1) ?;
      readReturn = read(fd, data + alreadyRead, 1);
    }
    
    alreadyRead++;
  }while(ends(data,responsecode));
  
  serialRead(data, readbytes);
  data[readbytes] = '\0';
  return &data[16];
}



void serialWrite(char *toWrite)
{
  int check = -1;
  while(check == -1){
    check = write(fd, toWrite, strlen(toWrite));
  }
}

void serialRead(char *data, int readbytes)
{
  int check;
  char tmp[readbytes];
  int readReturn = -1;
  while (readReturn == -1){
    readReturn = read(fd, &tmp, readbytes);
  }
  strcpy(data ,tmp);
}
